File: LeetCode/heap/TaskScheduler.java
package heap;

import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;

public class TaskScheduler {
    public int leastInterval(char[] tasks, int n) {
        int[] taskCounts = new int[26];
        for (char task : tasks) {
            taskCounts[task - 'A']++;
        }

        PriorityQueue<Integer> maxHeap = new PriorityQueue<>((a, b) -> b - a);
        for (int count : taskCounts) {
            if (count > 0) {
                maxHeap.offer(count);
            }
        }

        int time = 0;
        while (!maxHeap.isEmpty()) {
            int cycle = 0;
            List<Integer> cur = new ArrayList<>();
            for (int i = 0; i < n + 1; i++) {
                if (!maxHeap.isEmpty()) {
                    cur.add(maxHeap.poll());
                    cycle++;
                }
            }

            for (int count: cur) {
                count--;
                if (count > 0) {
                    maxHeap.offer(count);
                }
            }

            time += maxHeap.isEmpty() ? cycle : n + 1;
        }

        return time;
    }

    public static void main(String[] args) {
        TaskScheduler solution = new TaskScheduler();
        char[] tasks1 = {'A', 'A', 'A', 'B', 'B', 'B'};
        int n1 = 2;
        System.out.println(solution.leastInterval(tasks1, n1)); // Output: 8

        char[] tasks2 = {'A', 'C', 'A', 'B', 'D', 'B'};
        int n2 = 1;
        System.out.println(solution.leastInterval(tasks2, n2)); // Output: 6

        char[] tasks3 = {'A', 'A', 'A', 'B', 'B', 'B'};
        int n3 = 3;
        System.out.println(solution.leastInterval(tasks3, n3)); // Output: 10
    }
}

File: LeetCode_Python/heap/TaskScheduler.py
import heapq
from collections import Counter


class TaskScheduler:
    def leastInterval(self, tasks: list[str], n: int) -> int:
        count = Counter(tasks)

        heap = [-f for f in count.values() if f > 0]
        heapq.heapify(heap)

        res = 0

        while heap:
            cycle = n + 1
            remain = []
            task = 0
            while cycle > 0 and heap:
                cur_freq = -heapq.heappop(heap)
                if cur_freq > 1:
                    remain.append(-(cur_freq - 1))
                task += 1
                cycle -= 1

            for t in remain:
                heapq.heappush(heap, t)

            res += task if not heap else n + 1

        return res


if __name__ == "__main__":
    solution = TaskScheduler()

    tasks1 = ["A", "A", "A", "B", "B", "B"]
    n1 = 2
    print("Expected Output: 8")
    print("Actual Output:", solution.leastInterval(tasks1, n1))

    tasks2 = ["A", "A", "A", "B", "B", "B"]
    n2 = 0
    print("Expected Output: 6")
    print("Actual Output:", solution.leastInterval(tasks2, n2))

    tasks3 = ["A", "A", "A", "A", "A", "A", "B", "C", "D", "E", "F", "G"]
    n3 = 2
    print("Expected Output: 16")
    print("Actual Output:", solution.leastInterval(tasks3, n3))
